lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClassCheck.hs

```diff
@@ -90,27 +47,29 @@ runDeriveCheck mn mb = mconcat <$> traverse go (S.toList $ mbInstances mb)
           [] -> pure ()
           xs -> Left $ FailedToSolveConstraints mn xs i
 
--- ModuleBuilder is suitable codegen input,
--- and is (relatively) computationally expensive to
--- construct, so we return it here if successful.
-validateTypeClasses' :: P.CompilerInput -> Either TypeClassError (M.Map (P.InfoLess P.ModuleName) ModuleBuilder)
-validateTypeClasses' ci = do
-  -- detectSuperclassCycles ci
+runDeriveCheck' :: PC.CompilerInput -> Either TypeClassError (M.Map (PC.InfoLess PC.ModuleName) ModuleBuilder)
+runDeriveCheck' ci = do
   moduleBuilders <- mkBuilders ci
   void $ traverseWithKey runDeriveCheck moduleBuilders
   pure moduleBuilders
 
--- maybe use Control.Exception? Tho if we're not gonna catch it i guess this is fine
-validateTypeClasses :: P.CompilerInput -> IO (M.Map (P.InfoLess P.ModuleName) ModuleBuilder)
-validateTypeClasses ci = case validateTypeClasses' ci of
+runCheck :: PC.CompilerInput -> Maybe P.CompilerError
+runCheck ci = case Super.runCheck ci of
+  Left errs -> Just $ defMessage & P.tyClassCheckErrors .~ errs
+  Right () -> case runDeriveCheck' ci of
+    Left err -> Just $ defMessage & P.internalErrors .~ [defMessage & P.msg .~ ("TODO(bladyjoker): Use proper errors" <> (Text.pack . show $ err))]
+    Right _ -> Nothing
```

Comment:
Don't use `Maybe` to return an error - it really doesn't behave in the appropriate way for this. 
Use Either and return unit.

Reply:
Yeah you're right. On it